File: robotica_desplazamiento.py
import heapq

# Datos de costos representados en una matriz
costos = [
    [11.2, 10.2,  9.2,  8.2,  None, 6.8,  5.8,  4.8,  3.8,  3.4,  3.0],
    [10.8,  9.8,  8.8,  7.8,  6.8,  5.8,  4.8,  4.4,  None, 2.4,  2.0],
    [11.2, 10.2,  None, None, 6.4,  5.4,  4.4,  3.4,  None, 1.4,  1.0],
    [10.8,  9.8,  8.8,  7.8,  6.8, None, 4.0,  3.0,  2.0,  1.0,  0.0],
    [11.2, 10.2,  9.2,  8.2,  7.8, None, 4.4,  3.4,  2.4,  1.4,  1.0]
]

# Verifica si la posicion está dentro de los limites de la matriz
def es_valido(x, y):
    return 0 <= x < len(costos) and 0 <= y < len(costos[0]) and costos[x][y] is not None

# Obtiene los vecinos validos de una posicion
def obtener_vecinos(x, y):
    vecinos = [
        (x + 1, y),     # Abajo
        (x - 1, y),     # Arriba
        (x, y + 1),     # Derecha
        (x, y - 1),     # Izquierda
        (x + 1, y + 1), # Abajo-Derecha
        (x + 1, y - 1), # Abajo-Izquierda
        (x - 1, y + 1), # Arriba-Derecha
        (x - 1, y - 1)  # Arriba-Izquierda
    ]
    return [(nx, ny) for nx, ny in vecinos if es_valido(nx, ny)]

# Encuentra el mejor movimiento basado en el costo minimo
def mejor_movimiento(x, y):
    vecinos = obtener_vecinos(x, y)
    if vecinos:
        return min(vecinos, key=lambda pos: costos[pos[0]][pos[1]])
    return None

# Backtracking para encontrar el camino optimo hacia el objetivo (0)
def buscar_camino(x, y, camino):
    camino.append((x, y))

    # Condición de parada
    if costos[x][y] == 0.0:
        return True

    # Encontrar el mejor movimiento basado en el costo decreciente
    siguiente = mejor_movimiento(x, y)
    if siguiente:
        sx, sy = siguiente
        if costos[sx][sy] <= costos[x][y]:  # El costo debe ser decreciente
            if buscar_camino(sx, sy, camino):
                return True

    # Backtracking si no se encontro un camino valido
    camino.pop()
    return False

# Función principal para iniciar la busqueda desde (0, 0)
def encontrar_camino():
    camino = []
    if buscar_camino(0, 0, camino):
        return camino
    return None

# Mostrar el mapa
def mostrar_mapa():
    for fila in costos:
        print(" ".join(f"{costo:6.2f}" if costo is not None else " null " for costo in fila))

# Ejecucion del programa
if __name__ == "__main__":
    mostrar_mapa()
    print("\nBuscando el camino optimo desde (0, 0)...\n")
    camino = encontrar_camino()

    if camino:
        print("Camino encontrado:", camino)
    else:
        print("No se encontro un camino valido.")

File: robotica_desplazamiento.pl
% Representacion de casillas con su costo: costo (X, Y, Costo).
costo(0, 0, 11.2).
costo(0, 1, 10.2).
costo(0, 2, 9.2).
costo(0, 3, 8.2).
costo(0, 5, 6.8).
costo(0, 6, 5.8).
costo(0, 7, 4.8).
costo(0, 8, 3.8).
costo(0, 9, 3.4).
costo(0, 10, 3.0).

costo(1, 0, 10.8).
costo(1, 1, 9.8).
costo(1, 2, 8.8).
costo(1, 3, 7.8).
costo(1, 4, 6.8).
costo(1, 5, 5.8).
costo(1, 6, 4.8).
costo(1, 7, 4.4).
costo(1, 9, 2.4).
costo(1, 10, 2.0).

costo(2, 0, 11.2).
costo(2, 1, 10.2).
costo(2, 4, 6.4).
costo(2, 5, 5.4).
costo(2, 6, 4.4).
costo(2, 7, 3.4).
costo(2, 9, 1.4).
costo(2, 10, 1.0).

costo(3, 0, 10.8).
costo(3, 1, 9.8).
costo(3, 2, 8.8).
costo(3, 3, 7.8).
costo(3, 4, 6.8).
costo(3, 6, 4.0).
costo(3, 7, 3.0).
costo(3, 8, 2.0).
costo(3, 9, 1.0).
costo(3, 10, 0.0). % Objetivo

costo(4, 0, 11.2).
costo(4, 1, 10.2).
costo(4, 2, 9.2).
costo(4, 3, 8.2).
costo(4, 4, 7.8).
costo(4, 6, 4.4).
costo(4, 7, 3.4).
costo(4, 8, 2.4).
costo(4, 9, 1.4).
costo(4, 10, 1.0).

% Define movimiento valido y costo asociado
movimiento(X, Y, X1, Y1) :-
    (X1 is X + 1, X1 =< 4, Y1 is Y;
     X1 is X - 1, X1 >= 0, Y1 is Y;
     X1 is X, Y1 is Y + 1, Y1 =< 10;
     X1 is X, Y1 is Y - 1, Y1 >= 0;
     X1 is X + 1, X1 =< 4, Y1 is Y + 1, Y1 =< 10;
     X1 is X + 1, X1 =< 4, Y1 is Y - 1, Y1 >= 0;
     X1 is X - 1, X1 >= 0, Y1 is Y + 1, Y1 =< 10;
     X1 is X - 1, X1 >= 0, Y1 is Y - 1, Y1 >= 0),
    costo(X1, Y1, _).

% Encuentra el siguiente movimiento con el costo minimo
mejor_movimiento(X, Y, XMejor, YMejor) :-
    findall((Costo, X1, Y1), (
        movimiento(X, Y, X1, Y1),
        costo(X1, Y1, Costo),
        Costo >= 0
    ), Movimientos),
    sort(1, @=<, Movimientos, [(_, XMejor, YMejor)|_]).

% Busca el camino mas corto usando backtracking
buscar_camino(X, Y, [(X, Y)]) :-
    costo(X, Y, 0.0).  % Condición de parada: llegar al destino
buscar_camino(X, Y, [(X, Y)|Camino]) :-
    write('Visitando: '), write((X, Y)), nl,
    mejor_movimiento(X, Y, X1, Y1),
    buscar_camino(X1, Y1, Camino).

% Inicia la busqueda desde el punto (0,0)
encontrar_camino(Camino) :-
    buscar_camino(0, 0, Camino),
    write_term(Camino, [quoted(false), portray(true), max_depth(1000)]),
    nl.

% muestra el worldmap
mostrar_mapa :-
    forall(between(0, 4, X),
           (forall(between(0, 10, Y),
                   (costo(X, Y, Costo) ->
                       format('~4f ', [Costo]);
                       write('null  '))),
           nl)).

